### DISCUSSION TITLE
Collection import/export permissions

### DISCUSSION THREAD

**Discussion by boo-code** [directus/directus#12672] [Feature Requests]
There should be permission to enable/disable collections import/export. Most of the roles do not need such functionality at all, but export allows you to copy (steal) all the data very quickly. From security perspective it's not good, especially talking about customers data.

**Comment by rijkvanzanten** [comment]
Keep in mind that export/import only allows exporting/importing data you would otherwise have access to as well. There's no difference in paging over every api response or exporting it as a file 👍🏻 (as a matter of fact, the `?export` query parameter simply saves the current API call as a file, so there's no difference at all)

**Reply by boo-code** [reply]
Technically, you are right. It is always possible to take a photo of the screen just using your phone. In this way, data copy will be made too, and it's very hard to prevent it. But it's always about how complicated to make the copy and how much time does it take. To export 10 000 records in a second or to make 10 000 screen photos makes a difference. I do not like unlimited export in UI, and I do not like full data export in API or query just by adding ?export parameter.

The idea, that an employee or intern can make a copy of all the customers personal data from Directus based CRM in seconds sounds scary. If personal data is stolen, in such a case, an employee will get a few hundred euros fine (at worst), when the company will be fined for a full GDPR violation (fine can be up to €20 million or 4% of annual global turnover). Why? Because it will be a company's fault that it didn't protected personal data of customers.

This is the reason why full data export should be limited and controlled as much as possible. If I will have the possibility to limit queries to 20 results per request, I will use it. If I'll have the possibility to switch off export to file - I will use it. If I will have the possibility to switch off export UI - I will use it. Etc.

**Reply by rijkvanzanten** [reply]
But even that would be a false sense of security then wouldn't it? If you can query a maximum of 20 results per page, and can make the default 30 requests a second, you can still export those 10000 items in about 30 seconds with our without the export flag

**Reply by boo-code** [reply]
Technically, you are right again. But a lot of requests in a short period of time can be treated as an anomaly and a web firewall can be configured to block them. But web firewall is more like enterprise level security software. This is why, for SMBs would be good to have at least limitations at permission level. 

E.g. first level of protection. No export using Directus backend. It means, that it will not be enough to know how to click the mouse, and users will need to have some development skills at least.
The second level of protection. No export from API if requests are coming not from Directus backend. Requests limits per time frame can also be added here.

It will not protect the data from smart users like developers (they have direct access to DB in any way), but will create some obstacles for not so advanced users just from simple and dumb data copying. Now the only protection I can have from an export in a few mouse clicks is this in Project Settings:

#sidebar > div > div.v-item-group.sidebar-detail-group > div:nth-child(5) {display: none;}

:)

**Reply by boo-code** [reply]
Ok, I just came up with another idea. Usually, there is not so much sensitive data in the whole data stack. In such a case, it could be enough to mark a field as not exportable. After that, such field will not be exportable into a file or used as a column in a collection view. Its content may be visible only on the details page. It may be a quick solution. Even with an accessible export, it will not be possible to get all the sensitive information in one file in a few clicks of a mouse.

**Reply by sfarrowGR** [reply]
Agree with @boo-code - While technically savy users could over come this issue, the average user has no idea how the system works.  From an end user/client perspective, they may want this turned off.  I have several clients that would never allow this into production if someone who has view permissions for users can just click a button and get a spreadsheet with all the users data.  From a data protection standpoint, we are making it too easy.

**Reply by benhaynes** [reply]
Always worth looking into more options for controlling access to user I/O.

**Reply by hibem** [reply]
A Band-Aid with better compatibility to avoid hiding other menu items, such as Flow,
`div.sidebar-detail:has(i[data-icon="import_export"]) {
    display: none;
}
div.sidebar-detail i[data-icon="import_export"] {
    display: none;
}
div.sidebar-detail i[data-icon="import_export"] ~ * {
    display: none;
}`

**Comment by ponceleon** [comment]
I agree, this is super important to protect it, if it is possible to share a guide, even references of how it can be done and one follows the path gratefully.

Could you please confirm if by creating an extension I can do this.

https://docs.directus.io/extensions/creating-extensions.html

I have not yet created an extension for directu v.9 but I would dare to do it.

Thanks I really think this is an important feature.

**Comment by jvgadea** [comment]
i am agree too with @boo-code is necesary cut that option for a users that not needed

**Comment by AlaaL** [comment]
Yes this is a very important feature and i don't know if it can be done using an extension ?

**Reply by DanielBiegler** [reply]
Yeah, take a look over at https://github.com/br41nslug/directus-extension-block-endpoints

But of course at this point you're changing internal functionality and all bets are off.

Another option is to use a CSS selector in the custom CSS in the settings to hide the export function but like Rijk said this truly is just security by obscurity. Its a bandaid at best.

**Reply by AlaaL** [reply]
You are correct its just a bandaid, but the CSS option seems like a good choice until the feature implemented into Directus.

Thank you for the suggestions.

**Comment by NilsBaumgartner1994** [comment]
Any updates on this?
Realy important for us when deploying for multiple customers

**Comment by JugaadiTek** [comment]
There is a certain level of security in having the choice to not brandish the easy way to "cut & run" with the data.

The people who already have the knowledge to check the network requests, and care enough to look. Probably can't stop them anyways... Thats a hiring, trust, management and/or toxic work culture problem.

The people who do not know/not motivated to learn about the tech... 
If not given a button 95% will not go the extra mile to figure out how to do it, but could be tempted to with the button there.

**Comment by NilsBaumgartner1994** [comment]
> There is a certain level of security in having the choice to not brandish the easy way to "cut & run" with the data.
> 
> The people who already have the knowledge to check the network requests, and care enough to look. Probably can't stop them anyways... Thats a hiring, trust, management and/or toxic work culture problem.
> 
> The people who do not know/not motivated to learn about the tech... If not given a button 95% will not go the extra mile to figure out how to do it, but could be tempted to with the button there.

yes but:
- implementing by own may result in outdated or poor quality which brings other problems.
- Adding an „Advanced Mode Button“ which reveals such options to import permissions

**Comment by alepietrobon** [comment]
Upvote here. I agree that this would solve the concerns for 95% of users

**Comment by pazazdigital** [comment]
To add to this, we also would like the option to hide the import/export feature on Collections for certain site Admins, but keep for others, and ideally manage this via an Access Policy. 

Beyond the already discussed security implications, what is perahps more import is that we really don't want our allow our clients' junior staff a level of access where they can go importing 1000s of bad records into the database. 

An update to this discussion soon would be appreciated. Many thanks.

**Reply by rijkvanzanten** [reply]
My main concern before was that it's a bit "security through obfuscation" 🤔 Hiding/showing the button in the client is one thing, but that doesn't necessarily prevent a bad actor from using the import/export in the API then. In your use case, is it just a matter of removing the button from the UX to avoid mistakes, or would you imagine it to work as a security feature as well, effectively blocking any batch import/export from the APIs as well?

**Reply by devoluxbv** [reply]
> My main concern before was that it's a bit "security through obfuscation" 🤔 Hiding/showing the button in the client is one thing, but that doesn't necessarily prevent a bad actor from using the import/export in the API then. In your use case, is it just a matter of removing the button from the UX to avoid mistakes, or would you imagine it to work as a security feature as well, effectively blocking any batch import/export from the APIs as well?

You’re thinking like an IT person, focusing on APIs and technical aspects. However, this feature is designed to prevent a regular user from exporting all data with a single button.

Regular users often don’t even know what an API is. They use the interface to perform their job duties, such as managing invoice payments, calling clients from the database, filling out forms, closing deals, etc. Now imagine an employee is fired and decides to export the contact database “just in case.” Maybe they plan to call some clients themselves or even sell the data. In a single moment, your organization’s data could fall into the wrong hands.

Before you know it, you might start receiving calls from an Indian call center (just a joke, but you get the point).

It seems you’re focusing on Directus as a backend/CMS system for websites. (Even in this case, the ability to mass export content can be risky if I hire a freelancer just to write or edit a few posts directly in Directus.)

However, we use Directus as a CRM system, and we manage a very large customer database. Because of this, I prefer to both prevent certain users from using the export feature and have a way to track and see who exported what and when.

**Reply by rijkvanzanten** [reply]
> However, this feature is designed to prevent a regular user from exporting all data with a single button.

That really depends on who's the reader and what they're trying to achieve. There's folk out there that will assume that disabling export/import will prevent programmatic export/import as well. It's very important to be super clear on what's just toggling a button in the app versus a security measure.

> Regular users often don’t even know what an API is. They use the interface to perform their job duties, such as managing invoice payments, calling clients from the database, filling out forms, closing deals, etc.

Depends on what you're building and what your users are! 🙂 Your examples here — while super valid! — are also CRM specific. Directus is used for all sorts of use cases ranging from more traditional cms/crm use to managing video game levels, museum expositions, or airline traffic. There's quite a lot of use cases where a technical-user might disable exporting/importing thinking it prevents malicious data extraction from the platform, not knowing that the API is still a vulnerable target. That doesn't apply to _all_ users, but shouldn't be negated either.

> It seems you’re focusing on Directus as a backend/CMS system for websites. (Even in this case, the ability to mass export content can be risky if I hire a freelancer just to write or edit a few posts directly in Directus.)

Nope! I'm focusing on Directus as a builder for folks to make their ideal management systems in. A lot of folk use it for CMS-like use cases, but that's not the exclusive focus here 🙂 

> However, we use Directus as a CRM system, and we manage a very large customer database. Because of this, I prefer to both prevent certain users from using the export feature and have a way to track and see who exported what and when.

I think you're touching on a perfect example of why the whole security / api / technical aspect is so important. The _"have a way to track and see who exported what and when"_ is inherently a technical requirement. If we were to implement that as just a studio UI thing, there's an API angle that's not tracked and therefore the tracking can't be relied on for access logging security. 

I know this can quickly feel like bike shedding, but the devil is in the details and this type of minutiae can be a make it or break it moment for a thing like this 🙂

**Reply by pazazdigital** [reply]
@rijkvanzanten — I agree with original post here, and feel you are perhaps somewhat missing the point. We are currently building a CMS — and exploring using Directus — and have just tested as a User, who we had configured as a Editor (minimal priviledges), App only and no Admin rights. 

They were able to export all content they had access to to a csv and also import a random csv into a table, inserting 1000s of bad records.

What would be helpful is a separate set of permissions for importing/exporting data - covering both App or API.

**Reply by rijkvanzanten** [reply]
No no that is exactly the point I'm trying to make ha! 

> What would be helpful is a separate set of permissions for importing/exporting data - covering both App or API.

This is exactly what I'm getting at too 🙂 Earlier in this thread the suggestion was made just to hide/show the button on the app and that it wasn't important to consider the API piece as "regular users" don't know it even exists.

**Reply by pazazdigital** [reply]
I understand and agree we shoudn't just 'hide' buttons, but at present it proves a big operational security issue for us/our clients. 

So how do we stop App users importing/exporting data on-mass, with no checks/balances?

As I say, we have tested and a low-level website Editor was able to import 1000s of records (that didn't even match the structure of the table we had setup).

**Reply by pazazdigital** [reply]
ps. btw, it was one of the Directus' own senior team who suggested to me that we hide the buttons using custom CSS on a call last week.

**Comment by pazazdigital** [comment]
Hi. Ideally the import/export feature would be something that could be turned on/off for for both UX and/or API — but more pressing is the need to to show/hide it based on a user policy in the App (user policy already has no API access rights). Thanks.

**Reply by rijkvanzanten** [reply]
> but more pressing is the need to to show/hide it based on a user policy in the App (user policy already has no API access rights)

FWIW the permission policies you apply in the app control both the app and API permissions 👍🏻

**Reply by pazazdigital** [reply]
They don't allow us to stop bulk export or imports via the App. This is a posing some serious issues as we don't want Editors to have bulk import/export rights. Creating single entries and Editing only is what we require.

**Comment by hibem** [comment]
Hide the import/export button from the normal user in the browser, Settings->Appearance->Custom CSS
`div.sidebar-detail:has(i[data-icon="import_export"]) {
    display: none;
}`
for old browser:
`div.sidebar-detail i[data-icon="import_export"] {
    display: none;
}`
`div.sidebar-detail i[data-icon="import_export"] ~ * {
    display: none;
}`

**Comment by PetheanRaj-A-Mitrah** [comment]
Hi @boo-code 
I am using [Directus 11.15.4](https://github.com/directus/directus/releases)
Need to know how to handle the role-based export and export permissions in the sidebar?

because I need to show the import and export only to Admin.

**Comment by domdus** [comment]
Hey guys I just released this extension which covers hiding of import/export for roles:
https://github.com/domdus/directus-extension-module-permissions

"Module Permissions" in the marketplace.
Please try.
